lib/cog/command/pipeline/executor.ex

```diff
@@ -198,169 +181,107 @@ defmodule Cog.Command.Pipeline.Executor do
+        # TODO: what is this error case?
+        Helpers.send_error(msg, state.request, state.mq_conn)
         fail_pipeline(state, :binding_error, "Error preparing to execute command pipeline '#{state.request["text"]}': #{msg}")
-      bound ->
-        {current_bound, bound_scope} = collect_bound(bound)
-        {:next_state, :maybe_expand, %{state | current_bound: current_bound, scope: bound_scope}, 0}
-    end
-  end
-
-  @doc """
-  `expand` -> {:stop, :shutdown} | {:next_state, :get_options}
-
-  If a command begins with 'user' or 'site' it is treated as an alias and expanded.
-  """
-  def maybe_expand(:timeout, %__MODULE__{current: current_bound}=state) do
-    case AliasExpander.expand(current_bound.command, state.user) do
-      {:ok, :not_an_alias} ->
-        {:next_state, :get_options, state, 0}
-      {:ok, invocations} ->
-        [new_current|rest] = invocations
-        {:next_state, :bind, %{state | current: new_current, remaining: rest ++ state.remaining}, 0}
-      {:error, msg} ->
-        Helpers.send_reply(msg, state.request, state.mq_conn)
-        fail_pipeline(state, :alias_expansion_error, "Error expanding alias '#{current_bound.command}': #{msg}")
-    end
-  end
-
-  @doc """
-  `get_options` -> {:stop, :shutdown} | {:next_state, :check_permission}
-
-  Runs the option interpreter on the current bound invocation.
-  """
-  def get_options(:timeout, %__MODULE__{current_bound: current_bound}=state) do
-    case interpret_options(current_bound) do
-      {:ok, current_bound} ->
-        {:next_state, :maybe_enforce, %{state | current_bound: current_bound}, 0}
-      {:not_found, current_bound} ->
-        Helpers.send_idk(state.request, current_bound.command, state.mq_conn)
-        fail_pipeline(state, :option_interpreter_error, "Command '#{current_bound.command}' not found")
       error ->
         {:error, msg} = error
         Helpers.send_error(msg, state.request, state.mq_conn)
         fail_pipeline(state, :option_interpreter_error, "Error parsing options: #{inspect error}")
     end
   end
+  def plan_next_invocation(:timeout, %__MODULE__{invocations: [], output: [], destinations: destinations}=state) do
+    message = "Pipeline executed successfully, but no output was returned"
+    adapter = state.request["adapter"]
+    Enum.each(destinations, &publish_response(response_fn(message, adapter), &1, state))
+    {:stop, :shutdown, %{state | output: []}}
+  end
+  def plan_next_invocation(:timeout, %__MODULE__{invocations: [], output: output, destinations: destinations}=state) do
+    # No more invocations, but we've got data; run it through
+    # templating and send it out to all the destinations
+    command = state.current_plan.command
+    adapter = state.request["adapter"]
 
-  @doc """
-  `maybe_enforce` -> {:next_state, :maybe_collect_command} | {:next_state, :check_permission}
+    message = Enum.map_join(output, "\n", fn({data,template}) ->
+      render_template(command, adapter, template, data)
```

Comment:
I honestly don't know. I just seem to remember it coming up at some point. I imagined a scenario where a command would be invoked with some additional metadata; `:first`, `:last`. The command would then process it's input, append the results to it's collected outputs and return. If processing returns no results, for commands like filter, then the command just returns whatever outputs, if any, it's collected so far. When an input with `:first ` or `:last` is encountered we know to reset the output list. It would result in commands maintaining some state, but I don't think that would be terrible. And now that I think about it, we should be able to hide that bit from developers pretty easily if we need to.

Take what I say with a grain of salt though. I hadn't really thought about it. I'm totally making this up as I go based something that I may imagined us talking about :)